### DISCUSSION TITLE
ESP8266 - Does anything need to be flashed on it?

### DISCUSSION THREAD

**Discussion by isewise** [awawa-dev/HyperSPI#8] [Q&A] [answered]
Hello, I've been diving deep into the various aspects of this project, but Im a bit confused with the HyperSPI configuration. 

In regard to running HyperSPI from a RPI (In my case a RPI 3b+) to a ESP8266 (Mine: NodeMCU CP2102), does anything need to be flashed onto the ESP8266? I see in the Readme.md "Rpi acts as a master, ESP8266/ESP32 is in slave mode", but I wanted to double check. 

Thanks in advance.  

My planned hardware:
4K capture card connected via USB to RPI 3b+, powered by USB wall wart, running HyperHDR 18 connected to ESP8266 CP2102 via SPI cables; ESP8266, powered by USB charger (maybe TV USB if that works) connected to WS2815 (~300 LEDS, havent cut them to fit yet) via the data cable and ground, WS2815 powered by its own 12v 10amp power supply.

Thanks in advance.

**Comment by awawa-dev** [comment, marked as answer]
Hi
You definitely need to flash your ESP with HyperSPI firmware to make it work. You can download proper ESP8266 firmware for your LED strip type in the release section. On Rpi you just need HyperHDR.

**Reply by isewise** [reply]
Thanks. I don't know how I skipped over that. Too many tabs opened at once and skipping around too much. 

I tested my setup and it went well. 

FYI and for anyone else reading, BTF Lighting's WS2815 color order on Amazon is listed as GRB, but in HyperHDR I had to select RGB to get the correct colors.

**Reply by awawa-dev** [reply]
In HyperSPI for WS281x I'm using NeoGrbFeature feature to initialize the NeopixelBus library. It works for most such LED strips on the market and also supports correct LED strip color sequence when receiving RGB data at input. So the conversion from RGB to GRB is done on default in HyperSPI firmware driver.

**Reply by isewise** [reply]
Hello again and thanks for your assistant, 

I have another question regarding a capture card and latency. 

I have [this capture card](https://www.amazon.com/dp/B097DKNS1M?psc=1&ref=ppx_yo2ov_dt_b_product_details), which I though tdid 4k HDR but doesn't. I see you mention the EZCAP 320 a lot. Is that still recommended?
And finally, could this cheapo capture card be causing the latency issues I'm seeing?

**Reply by awawa-dev** [reply]
What's wrong with your grabber? Doesn't report to the video source/graphic card that is HDR capable? If only grabbers with a loop are to be considered: Ezcap 320 is powerful but HDR quality is not great, Ezcap 269 the quality is even worse, AV ACCESS 4KVC00 can simply disappear from the system if there is no video source  (so it must be present when HyperHDR starts) or for a moment if the resolution/refresh rate is changing (HyperHDR auto resume should help) but the quality is great...or my current grabber Elgato HD60x if you can afford it. All these grabbers don't support HDCP. Probably more practical and cheaper solution is HDMI splitter + solid MS2109 clone (avoid fake USB3.0 version). https://www.hyperhdr.eu/2022/04/usb-grabbers-hdr-to-sdr-quality-test.html 

For the grabber latency test: https://www.hyperhdr.eu/2021/10/usb-grabbers-grand-latency-test-part-i.html
Usually latency for HyperHDR is very low, under 70-80ms.
First thing is to disable smoothing and test it again.

**Reply by isewise** [reply]
> What's wrong with your grabber? Doesn't report to the video source/graphic card that is HDR capable?

Nope. My Nvidia Shield wont play HDR content from Plex and my PS4 didn't detect HDR support either. So I'm returning it. 

> If only grabbers with a loop are to be considered:

I'm not familiar with what a "a loop" is. Sorry.

> Probably more practical and cheaper solution is HDMI splitter + solid MS2109 clone 

If I understand this setup correctly, 4K HDR source --HDMI--> HDMI Splitter < Output A: MS2109 Clone --USB--> RPI; Output B: --HDMI-->TV. Is that correct? If so, that is a very easy solution. Any thoughts on this [card](https://www.amazon.com/Capture-Streaming-Broadcasting-Conference-Teaching/dp/B09FLN63B3/ref=sr_1_2?crid=19MMV336DHVXX&keywords=Rullz+MS2109&qid=1667695191&sprefix=rullz+ms2109%2Caps%2C204&sr=8-2)?

In the mean time, I do have my LG rooted, the piccap just doesn't reliably start on boot up, which is why I pursued the hardware focused solution.

**Reply by awawa-dev** [reply]
> I'm not familiar with what a "a loop" is. Sorry.

A grabber with built-in internal HDMI splitter, so you don't need an external one. Unfortunately they are usually limited and causing problems like yours because the splitter is not HDR-capable.

> If I understand this setup correctly, 4K HDR source --HDMI--> HDMI Splitter < Output A: MS2109 Clone --USB--> RPI; Output B: --HDMI-->TV. Is that correct? If so, that is a very easy solution. Any thoughts on this [card](https://www.amazon.com/Capture-Streaming-Broadcasting-Conference-Teaching/dp/B09FLN63B3/ref=sr_1_2?crid=19MMV336DHVXX&keywords=Rullz+MS2109&qid=1667695191&sprefix=rullz+ms2109%2Caps%2C204&sr=8-2)?

That grabber looks good. if Output A and Output B mean the output order then you want to connect a MS2109 clone to the output B (port 2), because the output A (port 1)  is usually prioritized by the HDMI splitter for higher-end devices like TV supports HDR. If you connect the MS2109 to the port 1, you probably won't have HDR.
